web/app.js

```diff
@@ -2674,18 +2702,66 @@ function webViewerWheel(evt) {
 
 function webViewerTouchStart(evt) {
   if (evt.touches.length > 1) {
-    // Disable touch-based zooming, because the entire UI bits gets zoomed and
-    // that doesn't look great. If we do want to have a good touch-based
-    // zooming experience, we need to implement smooth zoom capability (probably
-    // using a CSS transform for faster visual response, followed by async
-    // re-rendering at the final zoom level) and do gesture detection on the
-    // touchmove events to drive it. Or if we want to settle for a less good
-    // experience we can make the touchmove events drive the existing step-zoom
-    // behaviour that the ctrl+mousewheel path takes.
     evt.preventDefault();
+
+    if (evt.touches.length !== 2) {
+      PDFViewerApplication._touchInfo = null;
+      return;
+    }
+
+    const [touch0, touch1] = evt.touches;
+    PDFViewerApplication._touchInfo = {
+      distance:
+        Math.hypot(touch0.pageX - touch1.pageX, touch0.pageY - touch1.pageY) ||
+        1,
+    };
   }
 }
 
+function webViewerTouchMove(evt) {
+  if (!PDFViewerApplication._touchInfo || evt.touches.length !== 2) {
+    return;
+  }
+
+  evt.preventDefault();
+  const [touch0, touch1] = evt.touches;
+  const distance =
+    Math.hypot(touch0.pageX - touch1.pageX, touch0.pageY - touch1.pageY) || 1;
+  const previousDistance = PDFViewerApplication._touchInfo.distance;
+  const scaleFactor = distance / previousDistance;
+  if (Math.abs(scaleFactor - 1) < 1e-2) {
+    // Scale increase/decrease isn't significant enough.
+    return;
+  }
+  PDFViewerApplication._touchInfo.distance = distance;
+
+  if (PDFViewerApplication.supportsPinchToZoom) {
+    PDFViewerApplication.rescale(scaleFactor, {
+      drawingDelay: AppOptions.get("defaultZoomDelay"),
+    });
+    return;
+  }
+
+  const PIXELS_PER_LINE_SCALE = 30;
+  const delta = (distance - previousDistance) / PIXELS_PER_LINE_SCALE;
+  const ticks = PDFViewerApplication.accumulateWheelTicks(delta);
```

Comment:
Is re-using this generally correct, since a device with both a mouse and a touch-screen (e.g. a laptop) would then have one type of zooming effecting the other as far as I can tell?

Rather you should probably introduce a new method, e.g. named `accumulateTouchTicks` and being essentially a copy of the existing one, that you use instead here?